### DISCUSSION TITLE
How to call async def from sensor

### DISCUSSION THREAD

**Discussion by mqq-marek** [dagster-io/dagster#14700] [Q&A]
Sensor can not be defined as async def atm (Dagster 1.3.7) but I want to call async code from sensor (beanie).

How to make it in a proper way.

I try few approaches but none of them is working for me.

1. result = asyncio.run(to_await)
2.         # loop = asyncio.get_running_loop()
        loop = asyncio.get_event_loop()
        task = asyncio.run_coroutine_threadsafe(to_await, loop)
        result = task.result(timeout=5) 
3. loop = asyncio.new_event_loop()
    loop.run_forever()
    asyncio.set_event_loop(loop)
    t = Thread(target=background_loop, args=(loop,), daemon=True)
    t.start()
    task = asyncio.run_coroutine_threadsafe(to_await, loop)
    result = task.result(timeout=5)
---
def background_loop(loop: asyncio.AbstractEventLoop) -> None:
    asyncio.set_event_loop(loop)
    loop.run_forever()
4.     loop = asyncio.new_event_loop()
    loop.run_forever()
    asyncio.set_event_loop(loop)
    result = asyncio.run(to_await)

Any idea about proper way to do it with dagster?

**Comment by sryza** [comment]
Hi @mqq-marek - this is more of a general Python question than a Dagster question.  Here are a couple resources I found online on this subject:
- https://discuss.python.org/t/calling-coroutines-from-sync-code-2/24093
- https://stackoverflow.com/questions/44882586/calling-a-coroutine-synchronously

**Comment by jayhale-rp** [comment]
Considering that many sensors involve network requests, I'd really like to see `@sensor` be able to accept and `async def` instead of needing to establish a runtime within the sensor definition.

**Comment by azuline** [comment]
Same with asset checks and other coroutines. The solutions for calling an async function from a sync function are difficult to apply in `pytest_asyncio` tests where there is already an active event loop.
